pkg/ipcache/types.go

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ipcache
+
+import (
+	"github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+// prefixInfo holds all of the information (labels, etc.) about a given prefix
+// independently based on the ResourceID of the origin of that information, and
+// provides convenient accessors to consistently merge the stored information
+// to generate ipcache output based on a range of inputs.
+type prefixInfo map[types.ResourceID]*resourceInfo
+
+// resourceInfo is all of the information that has been collected from a given
+// resource (types.ResourceID) about this IP. Each field must have a 'zero'
+// value that indicates that it should be ignored for purposes of merging
+// multiple resourceInfo across multiple ResourceIDs together.
+type resourceInfo struct {
+	labels labels.Labels
+	source source.Source
+}
+
+// IPMetadata is an empty interface intended to inform developers using the
+// IPCache interface about which types are valid to be injected, and how to
+// update this code, in particular the merge(),unmerge(),isValid() methods
+// below.
+//
+// In an ideal world, we would use Constraints here but as of Go 1.18, these
+// cannot be used in conjunction with methods, which is how the information
+// gets injected into the IPCache.
+type IPMetadata interface{}
```

Comment:
nit: this could be `type IPMetadata any`

Reply:
One of my goals here is to move all of the "merging" code into this one file, right next to each other, so it can be easily compared side-by-side. If we start adding constraints onto the `IPMetadata` like requiring methods to merge/unmerge, then we will go back to a point where the code for merging / handling conflicts / resolving priority is spread across all of the callers elsewhere in the codebase. That makes it much easier to miss some newly introduced combination.

Reply:
Yeah that's true, having the callers implement their own methods to satisfy the interface wouldn't be great. I was thinking you could just warp/alias all the necessary types here, kind of like with labelMetadataImpl in my example, that way all your merge logic is in one place.

I only bring this up because I usually try to avoid type switching/asserting where possible (unless it's a container type). It kind of goes against using a static typing system imo.

Reply:
I see yeah, we could do that. But then if we end up warp/aliasing then it seems like we will need to end up doing the type switch/assert anyway?

I do think Go generics could eventually solve this for us by declaring it like this:

```
type IPMetadata interface{
    labels.Labels | foo | bar
}
```

However, I couldn't get this to work with the current implementation of go generics. Maybe at some point, but this is also not critical to solve this way.